File: internal/memory/migrations.go
package memory

import (
	"context"
	"database/sql"
	_ "embed"
	"fmt"
	"strings"
	"time"
)

//go:embed migrations/001_memory.sql
var schemaV1 string

//go:embed migrations/002_observation_title.sql
var schemaV2 string

//go:embed migrations/003_global_projects.sql
var schemaV3 string

//go:embed migrations/004_project_roots.sql
var schemaV4 string

//go:embed migrations/005_sdd.sql
var schemaV5 string

//go:embed migrations/006_sync.sql
var schemaV6 string

//go:embed migrations/007_local_write_versions.sql
var schemaV7 string

//go:embed migrations/008_sync_inbox_cursor_conflicts_bootstrap.sql
var schemaV8 string

//go:embed migrations/009_sync_push_results.sql
var schemaV9 string

//go:embed migrations/010_sync_outbox_claims.sql
var schemaV10 string

//go:embed migrations/011_sdd_ultra_plan.sql
var schemaV11 string

type migration struct {
	version                     int
	sql                         string
	requiresForeignKeysDisabled bool
}

var migrations = []migration{
	{version: 1, sql: schemaV1},
	{version: 2, sql: schemaV2},
	{version: 3, sql: schemaV3},
	{version: 4, sql: schemaV4},
	{version: 5, sql: schemaV5},
	{version: 6, sql: schemaV6},
	{version: 7, sql: schemaV7},
	{version: 8, sql: schemaV8},
	{version: 9, sql: schemaV9},
	{version: 10, sql: schemaV10},
	{version: 11, sql: schemaV11, requiresForeignKeysDisabled: true},
}

func applyMigrations(ctx context.Context, db *sql.DB, steps []migration) error {
	conn, err := db.Conn(ctx)
	if err != nil {
		return migrationError(ctx, "acquire connection", err)
	}
	defer conn.Close()
	if err := beginMigration(ctx, conn); err != nil {
		return err
	}
	inTransaction := true
	foreignKeysDisabled := false
	defer func() {
		if inTransaction {
			_, _ = conn.ExecContext(context.Background(), `ROLLBACK`)
		}
		if foreignKeysDisabled {
			_, _ = conn.ExecContext(context.Background(), `PRAGMA foreign_keys=ON`)
		}
	}()
	head, err := migrationHead(ctx, conn)
	if err != nil {
		return migrationError(ctx, "read migration ledger", err)
	}
	latest := 0
	for _, step := range steps {
		if step.version > latest {
			latest = step.version
		}
	}
	if head > latest {
		return fmt.Errorf("%w: database schema version %d is newer than supported version %d", ErrMigration, head, latest)
	}
	if requiresForeignKeysDisabled(steps, head) {
		if _, err = conn.ExecContext(ctx, `ROLLBACK`); err != nil {
			return migrationError(ctx, "release ownership for foreign-key rebuild", err)
		}
		inTransaction = false
		if _, err = conn.ExecContext(ctx, `PRAGMA foreign_keys=OFF`); err != nil {
			return migrationError(ctx, "disable foreign keys for rebuild", err)
		}
		foreignKeysDisabled = true
		if err = beginMigration(ctx, conn); err != nil {
			return err
		}
		inTransaction = true
		head, err = migrationHead(ctx, conn)
		if err != nil {
			return migrationError(ctx, "read migration ledger", err)
		}
		if head > latest {
			return fmt.Errorf("%w: database schema version %d is newer than supported version %d", ErrMigration, head, latest)
		}
	}
	for _, step := range steps {
		if step.version <= head {
			continue
		}
		_, err = conn.ExecContext(ctx, step.sql)
		if err == nil {
			_, err = conn.ExecContext(ctx, fmt.Sprintf(`PRAGMA user_version=%d`, step.version))
		}
		if err != nil {
			return migrationError(ctx, fmt.Sprintf("failed version %d", step.version), err)
		}
	}
	if foreignKeysDisabled {
		if err = foreignKeyCheck(ctx, conn); err != nil {
			return err
		}
	}
	if _, err = conn.ExecContext(ctx, `COMMIT`); err != nil {
		return migrationError(ctx, "commit", err)
	}
	inTransaction = false
	if foreignKeysDisabled {
		if err = enableForeignKeys(ctx, conn); err != nil {
			return err
		}
		foreignKeysDisabled = false
		if err = foreignKeyCheck(ctx, conn); err != nil {
			return err
		}
	}
	return nil
}

func beginMigration(ctx context.Context, conn *sql.Conn) error {
	for attempt := 0; ; attempt++ {
		_, err := conn.ExecContext(ctx, `BEGIN IMMEDIATE`)
		if err == nil {
			return nil
		}
		if err = waitForSQLite(ctx, attempt, err); err != nil {
			return migrationError(ctx, "acquire ownership", err)
		}
	}
}

func migrationHead(ctx context.Context, conn *sql.Conn) (int, error) {
	var head int
	err := conn.QueryRowContext(ctx, `PRAGMA user_version`).Scan(&head)
	return head, err
}

func requiresForeignKeysDisabled(steps []migration, head int) bool {
	for _, step := range steps {
		if step.version > head && step.requiresForeignKeysDisabled {
			return true
		}
	}
	return false
}

func enableForeignKeys(ctx context.Context, conn *sql.Conn) error {
	if _, err := conn.ExecContext(ctx, `PRAGMA foreign_keys=ON`); err != nil {
		return migrationError(ctx, "restore foreign keys", err)
	}
	var enabled int
	if err := conn.QueryRowContext(ctx, `PRAGMA foreign_keys`).Scan(&enabled); err != nil {
		return migrationError(ctx, "verify foreign keys enabled", err)
	}
	if enabled != 1 {
		return fmt.Errorf("%w: verify foreign keys enabled", ErrMigration)
	}
	return nil
}

func foreignKeyCheck(ctx context.Context, conn *sql.Conn) error {
	rows, err := conn.QueryContext(ctx, `PRAGMA foreign_key_check`)
	if err != nil {
		return migrationError(ctx, "verify foreign keys", err)
	}
	defer rows.Close()
	if rows.Next() {
		return fmt.Errorf("%w: verify foreign keys", ErrMigration)
	}
	if err = rows.Err(); err != nil {
		return migrationError(ctx, "verify foreign keys", err)
	}
	return nil
}

func waitForSQLite(ctx context.Context, attempt int, err error) error {
	locked := strings.Contains(strings.ToLower(err.Error()), "locked") || strings.Contains(strings.ToLower(err.Error()), "busy")
	if !locked || attempt >= 19 {
		return err
	}
	delay := time.Duration(attempt+1) * 10 * time.Millisecond
	if delay > 100*time.Millisecond {
		delay = 100 * time.Millisecond
	}
	timer := time.NewTimer(delay)
	defer timer.Stop()
	select {
	case <-ctx.Done():
		return ctx.Err()
	case <-timer.C:
		return nil
	}
}

func migrationError(ctx context.Context, operation string, _ error) error {
	if err := ctx.Err(); err != nil {
		return err
	}
	return fmt.Errorf("%w: %s", ErrMigration, operation)
}

File: packages/pi/src/sqlite/migrations.ts
import { createHash } from "node:crypto";
import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import type { DatabaseSync } from "node:sqlite";
import type { SQLiteDatabase } from "./node-sqlite.ts";
export type Migration = {
    version: number;
    sql: string;
    sha256: string;
    requiresForeignKeysDisabled?: boolean;
};
export const latestSchemaVersion = 23;
const resources = join(dirname(fileURLToPath(import.meta.url)), "../../resources/migrations");
type Manifest = {
    version: number;
    migrations: Array<{
        version: number;
        file: string;
        sha256: string;
    }>;
};
const manifest = JSON.parse(readFileSync(join(resources, "manifest.json"), "utf8")) as Manifest;
export const migrations: readonly Migration[] = Array.from({ length: latestSchemaVersion }, (_, index) => {
    const version = index + 1;
    const filename = `${String(version).padStart(3, "0")}_${["memory", "observation_title", "global_projects", "project_roots", "sdd", "sync", "local_write_versions", "sync_inbox_cursor_conflicts_bootstrap", "sync_push_results", "sync_outbox_claims", "sdd_ultra_plan", "sync_enrollment_recovery", "project_identities", "sync_portable_identities", "sync_portable_identity_adoptions", "sync_project_cursors", "sync_project_repairs", "sync_project_transitions", "sync_project_backup_intents", "provider_sessions", "provider_session_drafts", "observation_fts_metadata", "provider_session_leases"][index]}.sql`;
    const expected = manifest.migrations[index];
    if (!expected || manifest.version !== latestSchemaVersion || expected.version !== version || expected.file !== filename)
        throw new Error("invalid migration manifest");
    const sql = readFileSync(join(resources, filename), "utf8");
    const sha256 = createHash("sha256").update(sql).digest("hex");
    if (sha256 !== expected.sha256)
        throw new Error(`migration resource hash mismatch for version ${version}`);
    return { version, sql, sha256, requiresForeignKeysDisabled: version === 11 };
});
function raw(value: SQLiteDatabase | DatabaseSync): DatabaseSync { return "db" in value ? value.db : value; }
function version(db: DatabaseSync): number { return Number((db.prepare("PRAGMA user_version").get() as {
    user_version: number;
}).user_version); }
function foreignKeyCheck(db: DatabaseSync) { if (db.prepare("PRAGMA foreign_key_check").get() !== undefined)
    throw new Error("migration: foreign key check failed"); }
/** Matches the Go v11 rebuild algorithm: disable FK before BEGIN, then verify and restore it. */
export function applyMigrations(value: SQLiteDatabase | DatabaseSync, steps: readonly Migration[] = migrations): void {
    const db = raw(value);
    const latest = steps.reduce((head, step) => Math.max(head, step.version), 0);
    db.exec("BEGIN IMMEDIATE");
    let inTransaction = true;
    let rebuilding = false;
    try {
        let head = version(db);
        if (head > latest)
            throw new Error(`migration: database schema version ${head} is newer than supported version ${latest}`);
        rebuilding = steps.some((step) => step.version > head && step.requiresForeignKeysDisabled);
        if (rebuilding) {
            db.exec("ROLLBACK");
            inTransaction = false;
            db.exec("PRAGMA foreign_keys=OFF");
            db.exec("BEGIN IMMEDIATE");
            inTransaction = true;
            head = version(db);
            if (head > latest)
                throw new Error(`migration: database schema version ${head} is newer than supported version ${latest}`);
        }
        for (const step of steps)
            if (step.version > head) {
                db.exec(step.sql);
                db.exec(`PRAGMA user_version=${step.version}`);
            }
        if (rebuilding)
            foreignKeyCheck(db);
        db.exec("COMMIT");
        inTransaction = false;
        if (rebuilding) {
            db.exec("PRAGMA foreign_keys=ON");
            if (Number((db.prepare("PRAGMA foreign_keys").get() as {
                foreign_keys: number;
            }).foreign_keys) !== 1)
                throw new Error("migration: foreign keys not enabled");
            foreignKeyCheck(db);
        }
    }
    finally {
        if (inTransaction)
            try {
                db.exec("ROLLBACK");
            }
            catch { }
        if (rebuilding)
            try {
                db.exec("PRAGMA foreign_keys=ON");
            }
            catch { }
    }
}
